client/v2/client.go

```diff
@@ -0,0 +1,355 @@
+	SetPrecision(s string)
+
+	// Database returns the currently set database of this Batch
+	Database() string
+	// SetDatabase sets the database of this Batch
+	SetDatabase(s string)
+
+	// WriteConsistency returns the currently set write consistency of this Batch
+	WriteConsistency() string
+	// SetWriteConsistency sets the write consistency of this Batch
+	SetWriteConsistency(s string)
+
+	// RetentionPolicy returns the currently set retention policy of this Batch
+	RetentionPolicy() string
+	// SetRetentionPolicy sets the retention policy of this Batch
+	SetRetentionPolicy(s string)
+}
+
+// NewBatchPoints returns a BatchPoints interface based on the given config.
+func NewBatchPoints(c BatchPointsConfig) BatchPoints {
+	return &batchpoints{
+		database:         c.Database,
+		precision:        c.Precision,
+		retentionPolicy:  c.RetentionPolicy,
+		writeConsistency: c.WriteConsistency,
+	}
+}
+
+type batchpoints struct {
+	sync.Mutex
+
+	points           []*Point
+	database         string
+	precision        string
+	retentionPolicy  string
+	writeConsistency string
+}
+
+func (bp *batchpoints) AddPoint(p *Point) {
+	bp.Lock()
+	defer bp.Unlock()
+	bp.points = append(bp.points, p)
+}
+
+func (bp *batchpoints) Points() []*Point {
+	bp.Lock()
+	defer bp.Unlock()
+	return bp.points
+}
+
+func (bp *batchpoints) Precision() string {
+	return bp.precision
+}
+
+func (bp *batchpoints) Database() string {
+	return bp.database
+}
+
+func (bp *batchpoints) WriteConsistency() string {
+	return bp.writeConsistency
+}
+
+func (bp *batchpoints) RetentionPolicy() string {
+	return bp.retentionPolicy
+}
+
+func (bp *batchpoints) SetPrecision(p string) {
+	bp.precision = p
+}
+
+func (bp *batchpoints) SetDatabase(db string) {
+	bp.database = db
+}
+
+func (bp *batchpoints) SetWriteConsistency(wc string) {
+	bp.writeConsistency = wc
+}
+
+func (bp *batchpoints) SetRetentionPolicy(rp string) {
+	bp.retentionPolicy = rp
+}
+
+type Point struct {
+	pt models.Point
+}
+
+// NewPoint returns a point with the given timestamp
+func NewPoint(
+	name string,
+	tags map[string]string,
+	fields map[string]interface{},
+	t time.Time,
+) *Point {
+	return &Point{
+		pt: models.NewPoint(name, tags, fields, t),
+	}
+}
+
+// NewPointWithoutTime returns a point with a zero timestamp. This will be written
+// to the database without a timestamp, receiving one from InfluxDB whenever
+// it arrives.
+func NewPointWithoutTime(
+	name string,
+	tags map[string]string,
+	fields map[string]interface{},
+) *Point {
+	var t time.Time
+	return &Point{
+		pt: models.NewPoint(name, tags, fields, t),
+	}
+}
+
+// String returns a line-protocol string of the Point
+func (p *Point) String() string {
+	return p.pt.String()
+}
+
+// PrecisionString returns a line-protocol string of the Point, at precision
+func (p *Point) PrecisionString(precison string) string {
+	return p.pt.PrecisionString(precison)
+}
+
+func (c *client) Write(bp BatchPoints) error {
+	u := c.url
```

Comment:
Thanks for pointing that out, I've added more locks on all of the BatchPoints functions. There shouldn't need to be a lock here because the `Point()` and `AddPoint()` functions of `BatchPoints` are locked, preventing racy conflicts.